xarray/core/dataarray.py

```diff
@@ -158,6 +158,26 @@ def _infer_coords_and_dims(
     return new_coords, dims
 
 
+def _check_data_shape(data, coords, dims):
+    if data is dtypes.NA:
+        data = np.nan
+    if utils.is_scalar(data, test0d=False) and coords is not None:
+        if utils.is_dict_like(coords):
+            if dims is None:
+                return data
+            else:
+                data_shp = tuple(
+                    as_compatible_data(coords[k]).size if k in coords.keys() else 1
```

Comment:
You are right, I just tested `as_variable` and it works perfectly.

> It would be good to have a test or two covering coords being provided as tuple objects to be sure it works, though.

I am modifying a test to explicitly pass a list of tuples to `coords`.